docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,765 @@
+
+The request MUST be sent on the [partitioned topic][./10-waku-usage.md#partitioned-topic] of the sender.
+
+The payload of the response is:
+
+```protobuf
+message PushNotificationPreferencesResponse {
+  boolean success = 1;
+  ErrorType error = 2;
+  bytes request_id = 3;
+  PushNotificationPreferences preferences = 4;
+  string access_token = 5;
+
+  enum ErrorType {
+    UNKNOWN_ERROR_TYPE = 0;
+    MALFORMED_MESSAGE = 1;
+    VERSION_MISMATCH = 2;
+    UNSUPPORTED_TOKEN_TYPE = 3;
+    INTERNAL_ERROR = 4;
+  }
+}
+```
+
+[//]: (We can ratched and use a partitioned topic here)
+
+A client SHOULD listen for a response sent on their [partitioned topic][./10-waku-usage.md#partitioned-topic].
+
+If `success` is `true` the client has registered successfully.
+
+If `success` is `false`:
+
+- If `MALFORMED_MESSAGE` is returned, the request SHOULD not be retried without ensuring 
+  that is correctly formed.
+- If `INTERNAL_ERROR` is returned, the request MAY be retried, but the client MUST be 
+  backoff exponentially
+- If `VERSION_MISMATCH` is returned, the client SHOULD ensure that the remote version returned in `preferences` is integrated, and the version should be incremented and the request MAY be sent again.
+
+A client MAY register with multiple Push Notification Servers in order to increase availability.
+
+A client SHOULD make sure that all the notification services they registered with have the same information about their devices and tokens.
+
+If no response is returned the request SHOULD be considered failed and MAY be retried with the same server or a different one, but clients MUST exponentially backoff after each trial.
+
+If the request is successful a token is returned. This SHOULD be 
+[advertised](#advertising-a-push-notification-server) as described below if no public key filtering is necessary.
+
+## Re-registering with the push notification server
+
+A client SHOULD re-register with the node if the APN or FIREBASE token changes.
+
+When re-registering a client SHOULD ensure that it has the most up-to-date
+`PushNotificationPreferences`, update the part relative to their `installation_id`
+if necessary, increment `version` and send a `PushNotificationRegister` as described above.
+
+Once re-registered, a client SHOULD advertise the changes.
+
+
+## Changing options
+
+This is handled in exactly the same way as re-registering above.
+
+## Unregistering from push notifications
+
+To unregister a client MUST send a `PushNotificationRegister` request as described
+above with `unregister` in `PushNotificationPreferences` set to `true`, or removing
+their device information.
+
+The server MUST remove all data about this user if `unregistering` is `true`,
+apart from the `hash` of the public key and the `version` of the last options,
+in order to make sure that old messages are not processed.
+
+A client MAY unregister from a server on explicit logout if multiple accounts 
+are used on a single device.
```

Comment:
So I think this is a bit more of a UX concern,
but there are other things to consider that are implementation dependent, and I think `MAY` is best to give clients more freedom.

The best way to handle this is not to unregister only with the server on logout, because then you get into troubles if the user logs out when offline.

Say:
1) Go offline
2) Logout
3) Someone sends you a notification
4) You go online

The notification in 3 will be dispatched on IOS, there's no way around this as we don't have that kind of local control.

Ideally when you logout, you tell IOS that you don't want to receive notifications for that token, maybe by invalidating the previous one, but each client (android/IOS and maybe desktop) will kind of have to follow their own path. Some client might not even have a logout functionality (console client).

So I think MAY is more apt as it's very much related to the specific UX of a given client, but happy to change it to a SHOULD if you feel strongly about it.

Reply:
Thanks for the clarification. Makes sense.